lib/notifications/notification.ex

```diff
@@ -59,27 +60,34 @@ defmodule Notifications.Notification do
     :end_time
   ]
 
-  @spec create(t(), (DbNotification.t() -> t())) :: t()
-  def create(notification_without_id, insertion_callback \\ fn _ -> nil end) do
-    changeset =
-      DbNotification.changeset(
-        %DbNotification{},
-        Map.from_struct(notification_without_id)
-      )
+  @spec get_or_create(t(), (DbNotification.t() -> t())) :: t()
+  def get_or_create(notification_without_id, insertion_callback \\ fn _ -> nil end) do
+    identifying_fields =
+      Map.take(notification_without_id, [:start_time, :end_time, :block_id, :service_id, :reason])
+      |> Map.to_list()
 
-    id =
-      case insert(changeset) do
-        {:ok, db_notification} ->
+    {:ok, db_record} =
+      Skate.Repo.transaction(fn ->
+        existing_record = Skate.Repo.one(from(DbNotification, where: ^identifying_fields))
```

Comment:
Isn't there still a race condition here? It's inside a transaction but I don't think there's any guarantee that another client won't insert something between this `from` query and the `insert` below.

Reply:
I have just spent some time with the documentation and it looks like you're correct, will fix.